### PR TITLE
sysconfig/cloudinit.go: add "manual_cache_clean: true" to cloud-init restrict file

### PR DESCRIPTION
**Pull request by anonymouse64**
Without this key in the cloud-init restrict file, cloud-init on subsequent boots
will not know what instance-id to trust, as in multipass and other NoCloud
deployments the instance-id will only be available via the NoCloud CIDATA
filesystem label, but the snapd restriction file will not allow reading or
importing that filesystem label and so cloud-init will not know whether it is in
a new first boot like in the cloud where a "golden image" is booted with a given
cloud-init and then saved and rebooted with a new instance-id.

When cloud-init is unable to detect the instance-id of the current boot, it
assumes that it is booting a "golden image" and will search for new
configuration input and thus a new instance-id, but in the case of multipass,
where all cloud-init configuration is provided via NoCloud CIDATA filesystem
label drives, subsequent boots will "undo" the initial networking configuration
because cloud-init does not trust it's cached data and as such will do the
default configuration which overwrites the previous boot's cloud-init specified
configuration.

This additional key instructs cloud-init to trust it's cache of the instance-id,
thus not overwriting previous configuration.

Fixes: https://bugs.launchpad.net/snapd/+bug/1905983

Ideally I would also propose a nested regression test here, but I was unable to get the cloud-init configuration drive provided in the reproducer in the bug to work with cloud-init, so when I have a working version of that then I will propose separately a nested regression test for this new behavior. We did get confirmation that adding this key to zzzz_snapd.cfg is sufficient for multipass's usage of cloud-init though so I think the fix is verified to work.

### CODE REVIEW

**Review by mvo5 — APPROVED** [review]
Thank you

File: sysconfig/cloudinit.go [hunk 1]
```diff
@@ -151,9 +151,13 @@ var (
 	nocloudRestrictYaml = []byte(`datasource_list: [NoCloud]
 datasource:
   NoCloud:
-    fs_label: null`)
+    fs_label: null
+manual_cache_clean: true
```
Comment (bboozzoo) [inline sysconfig/cloudinit.go:155]: We should probably leave a reference to the bug itself and https://cloudinit.readthedocs.io/en/latest/topics/boot.html?highlight=manual_cache_clean#first-boot-determination right above `nocloudRestrictYaml`

**Review by bboozzoo — APPROVED** [review]

File: sysconfig/cloudinit.go [hunk 2]
```diff
@@ -151,9 +151,13 @@ var (
...
+`)
 
-	genericCloudRestrictYamlPattern = `datasource_list: [%s]`
+	genericCloudRestrictYamlPattern = `datasource_list: [%s]
+manual_cache_clean: true
```
Comment (pedronis) [inline sysconfig/cloudinit.go:159]: I'm not sure why we want do this for actual clouds,  the instance id there comes from the cloud itself, no?

**Review by pedronis — CHANGES_REQUESTED** [review]

File: sysconfig/cloudinit.go (hunk 2 above)
Comment (anonymouse64) [reply sysconfig/cloudinit.go]: As discussed this was removed and now `manual_cache_clean` is only used for the NoCloud datasource

File: sysconfig/cloudinit.go (hunk 1 above)
Comment (pedronis) [reply sysconfig/cloudinit.go:155]: agreed on both!

**Review by pedronis — APPROVED** [review]
thank you, but please follow Maciej suggestion to leave pointers to bug and flag documentation

File: sysconfig/cloudinit.go (hunk 1 above)
Comment (anonymouse64) [reply sysconfig/cloudinit.go:167]: added

**Comment by anonymouse64** [comment]
The failed tests here are all unrelated to this PR and this change, but are summarized here for posterity:
* google:ubuntu-16.04-32:tests/main/fake-netplan-apply which is fixed by https://github.com/snapcore/snapd/pull/9785
* google:ubuntu-16.04-64:tests/main/fake-netplan-apply which is fixed by https://github.com/snapcore/snapd/pull/9785
* google:ubuntu-20.04-64:tests/main/lxd:snapd_cgroup_just_outside due to the mysterious `ubuntu` line appearing in /etc/apt/sources.list in the nested container
* google:ubuntu-core-16-64:tests/core/netplan fixed by https://github.com/snapcore/snapd/pull/9785
* error preparing google:ubuntu-core-20-64:tests/main/ due to 408 from the store
* google-nested:ubuntu-20.10-64:tests/nested/classic/hotplug which is known but not debugged yet aiui
* google-nested:ubuntu-20.04-64:tests/nested/classic/hotplug which is same root cause as the groovy failure above I think
* google-nested:ubuntu-20.04-64:tests/nested/core20/basic which failed due to the fde-keys hook failure during seeding that was fixed by @mvo5 
* google-nested:ubuntu-20.04-64:tests/nested/core20/tpm again due to the fde-keys hook failure
* google-nested:ubuntu-20.04-64:tests/nested/manual/core20-early-config again due to the fde-keys hook failure
* google-nested:ubuntu-20.04-64:tests/nested/manual/core20-save:encrypted again due to the fde-keys hook failure
* google-nested:ubuntu-20.04-64:tests/nested/manual/grade-signed-cloud-init-testkeys again due to the fde-keys hook failure

**Pull request merged**